File: raytracer/Vector.java
package raytracer;

public class Vector {
	double x, y, z;
	public Vector(double x, double y, double z) {
		this.x = x;	
		this.y = y;
		this.z = z;
	}
	public Vector() {
		this.x = 0;
		this.y = 0;
		this.z = 0;
	}
	
	Vector cross(Vector a) {
		return new Vector(y * a.getZ() - z * a.getY(), z * a.getX() - x * a.getZ(), x * a.getY() - y * a.getX());
	}
	Vector plus(Vector a) {
		return new Vector(x + a.getX(), y + a.getY(), z + a.getZ());
	}
	double dot(Vector a) {
		return x * a.getX() + y * a.getY() + z * a.getZ();
	}
	Vector minus(Vector a) {
		return new Vector(x - a.getX(), y - a.getY(), z - a.getZ());
	}
	
	Vector times(double a) {
		return new Vector(a * x, a * y, a * z);
	}

	Vector divide(double a) {
		if (a == 0) {
			System.err.println("Divided by zero in " + this + ".divide()");
			return new Vector(0, 0, 0);
		}
		return this.times(1.0 / a);
	}
	
	Vector normalize() {
		double norm = Math.sqrt(x * x + y * y + z * z);
		if (norm == 0) {
			System.err.println("Divided by zero in " + this + ".normalize()");
			return new Vector(0, 0, 0);
		}
		return new Vector(x/norm, y/norm, z/norm);
	}
	
	public double getX() {
		return x;
	}

	public void setX(double x) {
		this.x = x;
	}

	public double getY() {
		return y;
	}

	public void setY(double y) {
		this.y = y;
	}

	public double getZ() {
		return z;
	}

	public void setZ(double z) {
		this.z = z;
	}
	
	public String toString() {
		return "((Vector) (" + x + "," + y + "," + z + "))";
	}
	
}

File: go/src/raytracer/geometry/vector.go
package geometry

import (
	"math"
	"math/rand"
)

type Vector struct {
	X float64
	Y float64
	Z float64
}

func BuildVector(elements []float64) Vector {
	return Vector{
		X: elements[0],
		Y: elements[1],
		Z: elements[2],
	}
}

var ZERO_VECTOR = BuildVector([]float64{0.0, 0.0, 0.0})

func (v1 Vector) Plus(v2 Vector) Vector {
	return BuildVector([]float64{v1.X + v2.X, v1.Y + v2.Y, v1.Z + v2.Z})
}

func (v1 Vector) Minus(v2 Vector) Vector {
	return BuildVector([]float64{v1.X - v2.X, v1.Y - v2.Y, v1.Z - v2.Z})
}

func (v1 Vector) Times(m float64) Vector {
	return BuildVector([]float64{v1.X * m, v1.Y * m, v1.Z * m})
}

func (v1 Vector) Vtimes(v2 Vector) Vector {
	return BuildVector([]float64{v1.X * v2.X, v1.Y * v2.Y, v1.Z * v2.Z})
}

func (v1 Vector) Dot(v2 Vector) float64 {
	return (v1.X * v2.X) + (v1.Y * v2.Y) + (v1.Z * v2.Z)
}

func (v Vector) Norm() float64 {
	return math.Sqrt(v.X * v.X + v.Y * v.Y + v.Z * v.Z);
}

func (v Vector) Normalize() Vector {
	norm := v.Norm()
	if norm == 0 { panic("Tried to normalize a zero vector") }

	return BuildVector([]float64{v.X / norm, v.Y / norm, v.Z / norm})
}

func (v1 Vector) Cross(v2 Vector) Vector {
	x := v1.Y * v2.Z - v1.Z * v2.Y
	y := v1.Z * v2.X - v1.X * v2.Z
	z := v1.X * v2.Y - v1.Y * v2.X

	return BuildVector([]float64{x, y, z})
}

func (v Vector) Abs() Vector {
	return BuildVector([]float64{math.Abs(v.X), math.Abs(v.Y), math.Abs(v.Z)})
}

func (v Vector) Contains(value float64) bool {
	return v.X == value || v.Y == value || v.Z == value
}

func (v Vector) Jitter() Vector {
	return BuildVector([]float64{randy(v.X), randy(v.Y), randy(v.Z)})
}

func randy(input float64) float64 {
	return input + (rand.Float64() / 5.0)
}
